internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -86,173 +89,277 @@ func (r *Reconciler) getAllMachineSetsAndSyncRevision(ctx context.Context, d *cl
-		r.recorder.Eventf(d, corev1.EventTypeNormal, "SuccessfulCreate", "Created MachineSet %q", newMS.Name)
+	// Construct the basic MachineSet.
+	desiredMS := &clusterv1.MachineSet{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: clusterv1.GroupVersion.String(),
+			Kind:       "MachineSet",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: deployment.Namespace,
+			// Note: By setting the ownerRef on creation we signal to the MachineSet controller that this is not a stand-alone MachineSet.
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(deployment, machineDeploymentKind)},
+			UID:             uid,
+			Finalizers:      finalizers,
+		},
+		Spec: clusterv1.MachineSetSpec{
+			Replicas:    &replicas,
+			ClusterName: deployment.Spec.ClusterName,
+			Template: clusterv1.MachineTemplateSpec{
+				Spec: machineTemplateSpec,
+			},
+		},
 	}
 
-	err = r.updateMachineDeployment(ctx, d, func(innerDeployment *clusterv1.MachineDeployment) {
-		mdutil.SetDeploymentRevision(d, newRevision)
-	})
+	// Set the in-place mutable fields.
+	// When we create a new MachineSet we will just create the MachineSet with those fields.
+	// When we update an existing MachineSet will we update the fields on the existing MachineSet (in-place mutate).
+
+	// Set labels and .spec.template.labels.
+	desiredMS.Labels = mdutil.CloneAndAddLabel(deployment.Spec.Template.Labels,
+		clusterv1.MachineDeploymentUniqueLabel, uniqueIdentifierLabelValue)
+	// Always set the MachineDeploymentNameLabel.
+	// Note: If a client tries to create a MachineDeployment without a selector, the MachineDeployment webhook
+	// will add this label automatically. But we want this label to always be present even if the MachineDeployment
+	// has a selector which doesn't include it. Therefore, we have to set it here explicitly.
+	desiredMS.Labels[clusterv1.MachineDeploymentNameLabel] = deployment.Name
+	desiredMS.Spec.Template.Labels = mdutil.CloneAndAddLabel(deployment.Spec.Template.Labels,
+		clusterv1.MachineDeploymentUniqueLabel, uniqueIdentifierLabelValue)
+
+	// Set selector.
+	desiredMS.Spec.Selector = *mdutil.CloneSelectorAndAddLabel(&deployment.Spec.Selector, clusterv1.MachineDeploymentUniqueLabel, uniqueIdentifierLabelValue)
+
+	// Set annotations and .spec.template.annotations.
+	if desiredMS.Annotations, err = mdutil.ComputeMachineSetAnnotations(log, deployment, oldMSs, existingMS); err != nil {
+		return nil, errors.Wrap(err, "failed to compute desired MachineSet: failed to compute annotations")
+	}
+	desiredMS.Spec.Template.Annotations = cloneStringMap(deployment.Spec.Template.Annotations)
+
+	// Set all other in-place mutable fields.
+	desiredMS.Spec.MinReadySeconds = pointer.Int32Deref(deployment.Spec.MinReadySeconds, 0)
+	desiredMS.Spec.DeletePolicy = pointer.StringDeref(deployment.Spec.Strategy.RollingUpdate.DeletePolicy, "")
+	desiredMS.Spec.Template.Spec.NodeDrainTimeout = deployment.Spec.Template.Spec.NodeDrainTimeout
```

Comment:
nit: might be useful to have in-place fields list in a constant so it's reused across code base e.g. machineTemplateDeepCopyIgnoringInPlacePropagatedFields

Reply:
Just for my understanding. How would that look like?

Today we have a bunch of place where we are setting in-place mutable fields. I'm not sure how we can use a const with in-place mutable fields across those places (assuming you don't mean that we should have a const with field paths and then use reflecting to read/write the fields)